hplip: Fix missing library

bkchr opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

bkchr committed:
hplip: Add missing library

Otherwise, things like `hp-setup` fail finding gi:

```
HP Linux Imaging and Printing System (ver. 3.20.5)
Printer/Fax Setup Utility ver. 9.0

Copyright (c) 2001-18 HP Development Company, LP
This software comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to distribute it
under certain conditions. See COPYING file for more details.

Traceback (most recent call last):
  File "/nix/store/rmzq8c4n47s4q4p31w18waz0zrba2ljl-hplip-3.20.5/share/hplip/setup.py", line 314, in <module>
    ui = import_module(ui_package + ".setupdialog")
  File "/nix/store/n0b076p351ma864q38if4yglsg99hw2s-python3-3.8.3/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/nix/store/rmzq8c4n47s4q4p31w18waz0zrba2ljl-hplip-3.20.5/share/hplip/ui5/setupdialog.py", line 31, in <module>
    from base import device, utils, models, pkit
  File "/nix/store/rmzq8c4n47s4q4p31w18waz0zrba2ljl-hplip-3.20.5/share/hplip/base/pkit.py", line 38, in <module>
    from gi import _gobject as gobject
ModuleNotFoundError: No module named 'gi'
```

bkchr commented:
CC @flokli

flokli commented:
@bkchr Thanks for the PR! Can you elaborate a bit in how this fails? I was actually able to build this, so I'm curious…

vcunat commented:
Like this?  https://hydra.nixos.org/build/122332330

flokli commented:
@vcunat no, it failed on master yesterday, and with https://github.com/NixOS/nixpkgs/pull/89120 merged, didn't. It currently builds on master.

bkchr commented:
Sorry, I should have made it clear. It builds, but it fails at running hp-setup for example. Other commands probably fail as well.

flokli committed:
hplip: use wrapQtApp

Otherwise, things like hp-setup segfault if you run them from another
nixpkgs checkout:

```
HP Linux Imaging and Printing System (ver. 3.20.5)
Printer/Fax Setup Utility ver. 9.0

Copyright (c) 2001-18 HP Development Company, LP
This software comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to distribute it
under certain conditions. See COPYING file for more details.

qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

[1]    4543 abort (core dumped)  result/bin/hp-setup
```

flokli commented:
Hmm, even with that patch, running the application (from another nixpkgs checkout) fails, as the QT stuff isn't wrapped:

```
HP Linux Imaging and Printing System (ver. 3.20.5)
Printer/Fax Setup Utility ver. 9.0

Copyright (c) 2001-18 HP Development Company, LP
This software comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to distribute it
under certain conditions. See COPYING file for more details.

qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

[1]    4543 abort (core dumped)  result/bin/hp-setup
```

flokli commented:
I updated the commit message to include some more detail, and pushed a second commit doing the qt wrapping for the hplip binaries. PTAL.

bkchr commented:
Hmm. @flokli you probably not use KDE? That is maybe the reason why I did not get the problems with QT.

flokli commented:
Nah, the problem can also happen in other cases, too: https://github.com/NixOS/nixpkgs/issues/65399

doronbehar commented on pkgs/misc/drivers/hplip/default.nix:
```diff
@@ -219,6 +222,10 @@ python3Packages.buildPythonApplication {
     remove-references-to -t ${stdenv.cc.cc} $(readlink -f $out/lib/*.so)
+  '' + stdenv.lib.optionalString withQt5 ''
+    for f in $out/bin/hp-*;do
+      wrapQtApp $f
```
FYI, I'm pretty sure the binaries are wrapped twice now. It should have been:

```
  preFixup = ''
    makeWrapperArgs+=("''${qtWrapperArgs[@]}")
  ''
```

But it's not that bad...

flokli replied:
Yeah, if someone wants to clean this up, I'd happily review a PR, but don't feel it a pressing enough issue to do it myself :-)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/unable-to-install-hp-printer-with-proprietary-driver-despite-setting-allowunfree-true-in-etc-nixos-configuration-nix/7741/15
